nixos/modules/config/system-path.nix

```diff
@@ -7,6 +7,118 @@ with lib;
+        };
+
+      drvToAttrsFolder = acc: drv: lib.recursiveUpdate acc {
+        "${builtins.unsafeDiscardStringContext drv.drvPath}" = drvToAttrs drv;
+      };
+
+      pkgsToAttrs = lib.foldl' drvToAttrsFolder {};
+
+      packageByOutPath =
+        let
+          mergeByOutPath = loc: defs:
+            if defs == [] then abort "This case should never happen."
+            # Return early if we only have one element
+            # This also makes it work for functions, because the foldl' below would try
+            # to compare the first element with itself, which is false for functions
+            else if length defs == 1 then (head defs).value
+            else (foldl' (first: def:
+              if def.value.outPath != first.value.outPath then
+                throw "The option `${options.showOption loc}' has conflicting definition values:${option.showDefs [ first.value.outPath def.value.outPath ]}"
+              else
+                first) (head defs) (tail defs)).value;
+        in
+        types.package // {
+          merge = loc: defs:
+            let res = mergeByOutPath loc defs;
+            in if builtins.isPath res || (builtins.isString res && ! builtins.hasContext res)
+              then toDerivation res
+              else res;
+        };
+    in
+    types.coercedTo
+      (types.listOf packageByOutPath)
+      pkgsToAttrs
+      (types.attrsOf (types.coercedTo
+        packageByOutPath
+        drvToAttrs
+        (types.submodule ( { config, lib, ... }: {
+          options = {
+            enable = (lib.mkEnableOption "the package") // {
+              default = true;
+            };
+            package = lib.mkOption {
+              type = packageByOutPath;
+              description = lib.mdDoc ''
+                The derivation that will be used for this package.
+              '';
+            };
+            outputs = lib.mkOption {
+              type = types.attrsOf types.bool;
+              description = lib.mdDoc ''
+                The outputs to install for this package.
+              '';
+            };
+          };
+
+          config = {
+            # We add the default outputs of the package by default.
+            # If a specific output has been selected, we only add that one.
+            outputs = mkOutputs (outputsFor config.package);
+          };
+        }))
+      ));
+
+  # This function takes the attribute set from the packages type and converts it
+  # back into a list that can be given to something like pkgs.buildEnv.
+  packagesApplyFun = ps:
+    let
+      enabled = lib.filterAttrs (_: p: p.enable) ps;
+      setOutputs = p:
+        let
+          outputsToInstall = lib.mapAttrsToList lib.const (lib.filterAttrs (_: lib.id) p.outputs);
+        in
+        lib.optional (lib.length outputsToInstall > 0) (p.package // {
+          meta = (p.package.meta or {}) // { inherit outputsToInstall; };
+        });
+    in
+    lib.flatten (lib.mapAttrsToList (_: setOutputs) enabled);
```

Comment:
`flatten` should never be necessary when you know the structure. In this case the faster `concatLists` can be used.